src/sql/src/query_model/dot.rs

```diff
@@ -48,26 +61,46 @@ enum DotLabel<'a> {
 /// The set of escaped characters is "|{}.
 struct DotLabelEscapedString<'a>(&'a str);
 
-impl DotGenerator {
-    fn new() -> Self {
+impl<'a> DotGenerator<'a> {
+    fn new(expr_humanizer: &'a dyn ExprHumanizer, with_types: bool) -> Self {
         Self {
             output: String::new(),
             indent: 0,
+            expr_humanizer,
+            with_types,
+        }
+    }
+
+    /// Derive attributes required for rendering the given [`Model`].
+    fn derive_required_attributes(&self, model: &mut Model, start_box: BoxId) {
+        // collect a set of attributes required by the given rules
+        let mut attributes = HashSet::new();
+        if self.with_types {
+            attributes.insert(Box::new(RelationType) as Box<dyn Attribute>);
         }
+        // add missing dependencies required to derive this set of attributes
+        transitive_closure(&mut attributes);
+        // order transitive closure topologically based on dependency order
+        let attributes = dependency_order(attributes);
```

Comment:
I see. Do `transitive_closure` and `dependency_order` ever run separately?

Reply:
No. I was thinking that we can introduce a dedicated `DerivableAttributes` type that can only be produced by a smart constructor that takes a list of attributes wraps the transitively closed version of that list in a dependency order. The argument of `derive_dft_attributes` will be changed to be a `DerivableAttributes` instance. This will ensure that we cannot pass an incomplete version of the attributes and we can hide `transitive_closure` and `dependency_order` inside the `DerivableAttributes` constructor.